File: src/utils/cache.ts
// Tiny disk cache: TTL-bound JSON entries, content-addressed by an md5 of the key.

import { mkdir } from "node:fs/promises";
import { join } from "node:path";
import { createHash } from "node:crypto";

export function hashKey(s: string): string {
  return createHash("md5").update(s).digest("hex");
}

interface Entry<T> {
  at: number;
  data: T;
}

/** Read a cached value if present and within `ttlMs`; otherwise null. */
export async function readCache<T>(dir: string, key: string, ttlMs: number): Promise<T | null> {
  try {
    const file = Bun.file(join(dir, hashKey(key) + ".json"));
    if (!(await file.exists())) return null;
    const entry = (await file.json()) as Entry<T>;
    if (Date.now() - entry.at > ttlMs) return null;
    return entry.data;
  } catch {
    return null;
  }
}

export async function writeCache<T>(dir: string, key: string, data: T): Promise<void> {
  try {
    await mkdir(dir, { recursive: true });
    const entry: Entry<T> = { at: Date.now(), data };
    await Bun.write(join(dir, hashKey(key) + ".json"), JSON.stringify(entry));
  } catch {
    // Caching is best-effort; never fail a request because the cache write failed.
  }
}

File: go-rewrite/internal/util/cache.go
// Tiny disk cache: TTL-bound JSON entries, content-addressed by an md5 of the
// key. Same envelope ({at: epoch-ms, data: …}) and filenames as the TS version,
// so both implementations share one cache.

package util

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

func hashKey(s string) string {
	sum := md5.Sum([]byte(s))
	return hex.EncodeToString(sum[:])
}

type cacheEntry struct {
	At   int64           `json:"at"`
	Data json.RawMessage `json:"data"`
}

// ReadCache reads a cached value into out if present and within ttl.
// Returns false on any miss or error.
func ReadCache(dir, key string, ttl time.Duration, out any) bool {
	raw, err := os.ReadFile(filepath.Join(dir, hashKey(key)+".json"))
	if err != nil {
		return false
	}
	var e cacheEntry
	if json.Unmarshal(raw, &e) != nil {
		return false
	}
	if time.Since(time.UnixMilli(e.At)) > ttl {
		return false
	}
	return json.Unmarshal(e.Data, out) == nil
}

// WriteCache stores a value. Best-effort — never fails the caller.
func WriteCache(dir, key string, data any) {
	raw, err := json.Marshal(data)
	if err != nil {
		return
	}
	e := cacheEntry{At: time.Now().UnixMilli(), Data: raw}
	blob, err := json.Marshal(e)
	if err != nil {
		return
	}
	if os.MkdirAll(dir, 0o755) != nil {
		return
	}
	_ = os.WriteFile(filepath.Join(dir, hashKey(key)+".json"), blob, 0o644)
}
